vcr/statuslist/store.go

```diff
@@ -0,0 +1,287 @@
+
+type sqlStore struct {
+	db *gorm.DB
+}
+
+// DB creates a new Session with the provided context.
+func (s *sqlStore) DB(ctx context.Context) *gorm.DB {
+	return s.db.WithContext(ctx)
+}
+
+func NewStatusListStore(db *gorm.DB) (*sqlStore, error) {
+	return &sqlStore{db: db}, nil
+}
+
+func (s *sqlStore) CredentialSubject(ctx context.Context, issuer did.DID, page int) (*credential.StatusList2021CredentialSubject, error) {
+	statusListCredential, err := toStatusListCredential(issuer, page)
+	if err != nil {
+		return nil, err
+	}
+
+	// confirm that status list credential exists
+	// SELECT * FROM status_list_credentials WHERE status_list_credential = 'statusListCredential'
+	err = s.DB(ctx).First(new(statusListCredentialRecord), "status_list_credential = ?", statusListCredential).Error
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			// status list credential does not exist or is not managed byt his node
+			return nil, errNotFound
+		}
+		return nil, err
+	}
+
+	// get revocations
+	var revocations []revocationRecord
+	// SELECT status_list_index FROM status_list_statuses WHERE status_list_credential = 'statusListCredential'
+	err = s.DB(ctx).Select("status_list_index").Find(&revocations, "status_list_credential = ?", statusListCredential).Error
+	if err != nil && errors.Is(err, gorm.ErrRecordNotFound) {
+		return nil, err
+	}
+
+	// make encodedList
+	bitstring := statuslist2021.NewBitstring()
+	for _, rev := range revocations {
+		if err = bitstring.SetBit(rev.StatusListIndex, true); err != nil {
+			// can't happen
+			return nil, err
+		}
+	}
+	encodedList, err := statuslist2021.Compress(*bitstring)
+	if err != nil {
+		// can't happen
+		return nil, err
+	}
+
+	// return credential subject
+	return &credential.StatusList2021CredentialSubject{
+		Id:            statusListCredential,
+		Type:          credential.StatusList2021CredentialSubjectType,
+		StatusPurpose: StatusPurposeRevocation,
+		EncodedList:   encodedList,
+	}, nil
+}
+
+func (s *sqlStore) Create(ctx context.Context, issuer did.DID, purpose StatusPurpose) (*credential.StatusList2021Entry, error) {
+	if purpose != StatusPurposeRevocation {
+		return nil, errUnsupportedPurpose
+	}
+
+	var credentialRecord statusListCredentialRecord
+
+	// optimistic locking to allocate a StatusListIndex for the issuer
+	for {
+		// Get last page for issuer
+		// SELECT * FROM status_list_credentials WHERE issuer = ? ORDER BY page DESC LIMIT 1;
+		tx := s.DB(ctx).Order("page DESC").First(&credentialRecord, "issuer = ?", issuer.String())
+		if tx.Error != nil {
+			if !errors.Is(tx.Error, gorm.ErrRecordNotFound) {
+				return nil, tx.Error
+			}
+
+			// first time issuer; prepare to create a new Page / StatusListCredential
+			credentialRecord = statusListCredentialRecord{
+				Issuer:          issuer.String(),
+				LastIssuedIndex: statuslist2021.MaxBitstringIndex,
+			}
+		}
+
+		// next index
+		credentialRecord.LastIssuedIndex++
```

Comment:
base 1 or 0? Because now the first index of the first page is 1 instead of 0?

Edit: oh now I see, you initialize `LastIssuedIndex: statuslist2021.MaxBitstringIndex,`, so it gets recognized as "full" page, so a new one gets created. THat's quite confusing?

Reply:
other options I though of all resulted in more code / separate code path for a new issuer. I added a comment to explain he behavior.